File: programmers/level1/완주하지못한선수.java
import java.util.*;
/*
* HashSet을 사용하면 동명이인을 처리할 수 없기 때문에
* HashMap을 사용하자!
*/
class Solution {
    public String solution(String[] participant, String[] completion) {
        String answer = "";
        
        Map<String, Integer> hashmap = new HashMap<>();
        for(String par : participant) {
            if(hashmap.containsKey(par)) hashmap.put(par, hashmap.get(par)+1);
            else hashmap.put(par, 1);
        }
        
        for(String com : completion) {
            if(hashmap.containsKey(com)) hashmap.put(com, hashmap.get(com)-1);
        }
        
        for(String str : hashmap.keySet()) {
            if(hashmap.get(str)!=0) answer += str;
        }
        
        return answer;
    }
}

File: tip/cpp/Map/완주하지못한선수.cpp
#include <string>
#include <vector>
#include <map>
#include <iostream>
#include <cstdio>

using namespace std;

string solution(vector<string> participant, vector<string> completion) {
    string answer = "";

    //(1) Map 선언하기
    map<string,int> map;

    //(2) 참가자들 Map에 <이름, 몇 명인지> 담기
    for(string name : participant) {
        map[name]++;
    }

    //(3) 참가자 중에서 완주한 사람 숫자 빼기
    for(string name : completion) {
        map[name]--;
    }
    
    //(4) 완주하지 못한 사람은?
    for(string name : participant){
        if(map[name]>0) {
            answer = name;
        }
    }

    return answer;
}
